common/src/main/java/org/conscrypt/TrustManagerImpl.java

```diff
@@ -962,6 +965,59 @@ private TrustAnchor findTrustAnchorBySubjectAndPublicKey(X509Certificate cert) {
         return (acceptedIssuers != null) ? acceptedIssuers.clone() : acceptedIssuers(rootKeyStore);
     }
 
+    /**
+     * Set the default hostname verifier that will be used for HTTPS endpoint identification.  If
+     * {@code null} (the default), endpoint identification will use the default hostname verifier
+     * set in {@link HttpsURLConnection#setDefaultHostnameVerifier(javax.net.ssl.HostnameVerifier)}.
+     */
+    synchronized static void setDefaultHostnameVerifier(HostnameVerifier verifier) {
+        defaultHostnameVerifier = verifier;
+    }
+
+    /**
+     * Returns the currently-set default hostname verifier.
+     *
+     * @see #setDefaultHostnameVerifier(HostnameVerifier)
+     */
+    synchronized static HostnameVerifier getDefaultHostnameVerifier() {
+        return defaultHostnameVerifier;
+    }
+
+    /**
+     * Set the hostname verifier that will be used for HTTPS endpoint identification.  If
+     * {@code null} (the default), endpoint identification will use the default hostname verifier
+     * set in {@link #setDefaultHostnameVerifier(HostnameVerifier)}.
+     */
+    void setHostnameVerifier(HostnameVerifier verifier) {
+        this.hostnameVerifier = verifier;
+    }
+
+    /**
+     * Returns the currently-set hostname verifier for this instance.
+     *
+     * @see #setHostnameVerifier(HostnameVerifier)
+     */
+    HostnameVerifier getHostnameVerifier() {
+        return hostnameVerifier;
+    }
+
+    private HostnameVerifier getHttpsVerifier() {
+        if (hostnameVerifier != null) {
+            return hostnameVerifier;
+        }
+        HostnameVerifier defaultVerifier = getDefaultHostnameVerifier();
+        if (defaultVerifier != null) {
+            return defaultVerifier;
+        }
+        final javax.net.ssl.HostnameVerifier httpsUrlVerifier =
+            HttpsURLConnection.getDefaultHostnameVerifier();
+        return new HostnameVerifier() {
+            @Override public boolean verify(String hostname, SSLSession session) {
+                return httpsUrlVerifier.verify(hostname, session);
+            }
+        };
```

Comment:
nit: Might as well define an internal singleton rather than creating a new HostnameVerifier for every call:

```java
private enum DefaultHttpsUrlVerifier implements HostnameVerifier {
    INSTANCE;

    @Override public boolean verify(String hostname, SSLSession session) {
        final javax.net.ssl.HostnameVerifier httpsUrlVerifier = HttpsURLConnection.getDefaultHostnameVerifier();
        return httpsUrlVerifier.verify(hostname, session);
    }
}
```

Reply:
Good idea, done.